zookeeper/src/main/java/com/linecorp/armeria/common/zookeeper/ServerSetsInstance.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+// =================================================================================================
+// Copyright 2011 Twitter, Inc.
+// -------------------------------------------------------------------------------------------------
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this work except in compliance with the License.
+// You may obtain a copy of the License in the LICENSE file, or at:
+//
+//  https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// =================================================================================================
+package com.linecorp.armeria.common.zookeeper;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.Map;
+
+import javax.annotation.Nullable;
+
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.fasterxml.jackson.databind.annotation.JsonSerialize;
+import com.google.common.base.MoreObjects;
+import com.google.common.base.Objects;
+import com.google.common.collect.ImmutableMap;
+
+import com.linecorp.armeria.client.Endpoint;
+import com.linecorp.armeria.common.zookeeper.ServerSetsInstanceConverter.FinagleServiceInstanceDeserializer;
+import com.linecorp.armeria.common.zookeeper.ServerSetsInstanceConverter.FinagleServiceInstanceSerializer;
+
+/**
+ * A class that represents a service instance used by
+ * <a href="https://github.com/twitter/finagle/tree/develop/finagle-serversets">Finagle ServerSets</a>.
+ */
+@JsonSerialize(using = FinagleServiceInstanceSerializer.class)
+@JsonDeserialize(using = FinagleServiceInstanceDeserializer.class)
+public final class ServerSetsInstance {
+
+    // IDL is defined in https://github.com/twitter/finagle/blob/finagle-20.5.0/finagle-serversets/src/main/thrift/com/twitter/thrift/endpoint.thrift
+
+    @Nullable
+    private final Endpoint serviceEndpoint;
```

Comment:
Yeah, but I make it nullable because a user might not want to specify an endpoint when building `ZooKeeperRegistrationSpec`. The endpoint is automatically created using the default information.

Reply:
I understand what you are intended. `ServerSetsInstance` is used when building `ZooKeeperRegistrationSpec`. It could reduce duplicate code.
However, in my opinion, it seems not to correct usage of that. 
Because it is a public API and a user could implement and inject their own `ServerSetsInstance` converter via `ZooKeeperDiscoverySpec.serverSets(). If `serviceEndpoint` is nullable, he or she would think that nullable should be handled.

Reply:
> A user could implement and inject their own ServerSetsInstance converter via ZooKeeperDiscoverySpec.serverSets(). If serviceEndpoint` is nullable, he or she would think that nullable should be handled.

In this case, a use can check if the `serviceEndpoint` is null or not?
Because the getter annotated with `@Nullable`?

Reply:
> ServerSetsInstance is designed to transfer instance information.

Yes, but it's only used in the converter.
And because it's nullable, a user can check and just return null endpoint. Isn't that right?

Reply:
Yeah, that is possible. It is not a crucial point. I just thought this entity does not express the actual protocol. Because the response of `serviceEndpoint` from the ZooKeeper is always not null. So I think it is better to make it not null so that users don't need to check nullable.

Reply:
Had chat, I agreed with leaving with as it is.
Because `@Nullable` is no harm for users and we can reduce duplicate code.